### PR TITLE
Fix uneven head sequence parallelism bug (#6774)

### PR DESCRIPTION
**Pull request by Eugene29**
Here `gather_idx < 2` represents `is_first_all2all`. During the first all2all, `uneven_head_all2all` will be called if either `num_heads % seq_world_size != 0` or `get_num_kv_heads() is None`. 

During the second all2all, it'll return return `uneven_head_all2all` if and only if `get_num_kv_heads() is None` which is always set during the first uneven all2all. This means that there will no longer be issue where `uneven_head_all2all ` is returned for the second all2all because of `num_heads % seq_world_size != 0`.

Fixes: #6774

### CODE REVIEW

**Comment by Eugene29** [comment]
Fix #6774

**Comment by inkcherry** [comment]
LGTM， Many thanks!

File: deepspeed/sequence/layer.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def single_all_to_all(input, scatter_idx, gather_idx, batch_dim_idx, group, asyn
     # we only need num_heads once
     num_heads = input.shape[2]
 
-    if get_num_kv_heads() is not None or num_heads % seq_world_size != 0:
+    if get_num_kv_heads() is not None or (num_heads % seq_world_size != 0 and gather_idx < 2):
```
Comment (inkcherry) [inline deepspeed/sequence/layer.py:158]: By the way, can we use scatter_idx for the judgment here? This way, we can stay consistent with other parts by uniformly using scatter_idx to determine whether the current state is ```total_head``` or ```total_seq_len```

Comment (samadejacobs) [reply deepspeed/sequence/layer.py:158]: Ditto @inkcherry! Future refactor will have a better naming to differentiate between N/p (pre/post-attention) and h/p (attention block)

Comment (Eugene29) [reply deepspeed/sequence/layer.py:158]: Hi I agree, scatter_idx booleans are confusing. Maybe I can share my opinion on the refactoring on Wed if thats ok.

Comment (Eugene29) [reply deepspeed/sequence/layer.py]: Hey, I just made a quick change to use `scatter_idx`.

**Review by loadams — APPROVED** [review]

**Pull request merged**